packages/plugin-ext-headless/src/hosted/node/plugin-host-headless-rpc.ts

```diff
@@ -0,0 +1,94 @@
+// *****************************************************************************
+// Copyright (C) 2023 EclipseSource and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0.
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import { dynamicRequire } from '@theia/core/lib/node/dynamic-require';
+import { ContainerModule, injectable, inject } from '@theia/core/shared/inversify';
+import { EnvExtImpl } from '@theia/plugin-ext/lib/plugin/env';
+import { LocalizationExt } from '@theia/plugin-ext';
+import { LocalizationExtImpl } from '@theia/plugin-ext/lib/plugin/localization-ext';
+import { HEADLESSMAIN_RPC_CONTEXT } from '../../common/headless-plugin-rpc';
+import { HeadlessPluginManagerExtImpl } from '../../plugin/headless-plugin-manager';
+import { AbstractPluginHostRPC, ExtInterfaces } from '@theia/plugin-ext/lib/hosted/node/plugin-host-rpc';
+import { PluginHost } from '@theia/plugin-ext/lib/plugin/plugin-manager';
+import { PluginModel } from '@theia/plugin-ext/lib/common/plugin-protocol';
+import { ExtPluginApi, ExtPluginApiHeadlessInitializationFn } from '../../common/plugin-ext-headless-api-contribution';
+import { RPCProtocol } from '@theia/plugin-ext/lib/common/rpc-protocol';
+
+type HeadlessExtInterfaces = Pick<ExtInterfaces, 'envExt'|'localizationExt'>;
+
+/**
+ * The RPC handler for headless plugins.
+ */
+@injectable()
+export class HeadlessPluginHostRPC extends AbstractPluginHostRPC<HeadlessPluginManagerExtImpl, undefined, HeadlessExtInterfaces> {
+    @inject(EnvExtImpl)
+    protected readonly envExt: EnvExtImpl;
+
+    @inject(LocalizationExt)
+    protected readonly localizationExt: LocalizationExtImpl;
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    constructor(@inject(RPCProtocol) _rpc: RPCProtocol) {
+        super(
+            _rpc, 'HEADLESSPLUGIN_HOST', undefined,
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            (pluginHost: PluginHost, extInterfaces: HeadlessExtInterfaces, rpc: any) => {
+                const { envExt, localizationExt } = extInterfaces;
+                return new HeadlessPluginManagerExtImpl(pluginHost,
```

Comment:
Since we already have dependency injection, I wonder whether we should really hardcode the construction here. Maybe we can introduce a factory function or something similar so it could be overridden? Or is that not possible anyway since there is not much we can actually override at this point? Just wondering.